### PR TITLE
Kotlin 2.0.20 support

### LINKED ISSUE #1825: Kotlin 2.0.20 support
### How frequently does the bug occur?

Always

### Description

'2.1.0' fails with the below:

### Stacktrace & log output

```shell
Task :app:compileDebugKotlin FAILED
e: java.lang.NoSuchMethodError: 'java.lang.RuntimeException org.jetbrains.kotlin.backend.common.CompilationExceptionKt.wrapWithCompilationException(java.lang.Throwable, java.lang.String, org.jetbrains.kotlin.ir.declarations.IrFile, org.jetbrains.kotlin.ir.IrElement)'
	at io.realm.kotlin.compiler.RealmModelLowering.lowerFromModuleFragment(RealmModelLoweringExtension.kt:83)
	at io.realm.kotlin.compiler.RealmModelLoweringExtension.generate(RealmModelLoweringExtension.kt:53)
	at org.jetbrains.kotlin.fir.pipeline.ConvertToIrKt.applyIrGenerationExtensions(convertToIr.kt:437)
	at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.runActualizationPipeline(convertToIr.kt:246)
	at org.jetbrains.kotlin.fir.pipeline.Fir2IrPipeline.convertToIrAndActualize(convertToIr.kt:130)
	at org.jetbrains.kotlin.fir.pipeline.ConvertToIrKt.convertToIrAndActualize(convertToIr.kt:99)
	at org.jetbrains.kotlin.fir.pipeline.ConvertToIrKt.convertToIrAndActualize$default(convertToIr.kt:72)
	at org.jetbrains.kotlin.cli.jvm.compiler.pipeline.JvmCompilerPipelineKt.convertToIrAndActualizeForJvm(jvmCompilerPipeline.kt:196)
	at org.jetbrains.kotlin.cli.jvm.compiler.pipeline.JvmCompilerPipelineKt.convertAnalyzedFirToIr(jvmCompilerPipeline.kt:169)
	at org.jetbrains.kotlin.cli.jvm.compiler.pipeline.JvmCompilerPipelineKt.compileModulesUsingFrontendIrAndLightTree(jvmCompilerPipeline.kt:140)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:148)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:43)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:103)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:49)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:101)
	at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:464)
	at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:73)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.doCompile(IncrementalCompilerRunner.kt:506)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl(IncrementalCompilerRunner.kt:423)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileNonIncrementally(IncrementalCompilerRunner.kt:301)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compile(IncrementalCompilerRunner.kt:129)
	at org.jetbrains.kotlin.daemon.CompileServiceImplBase.execIncrementalCompiler(CompileServiceImpl.kt:675)
	at org.jetbrains.kotlin.daemon.CompileServiceImplBase.access$execIncrementalCompiler(CompileServiceImpl.kt:92)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:1660)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(Unknown Source)
	at java.rmi/sun.rmi.transport.Transport$1.run(Unknown Source)
	at java.rmi/sun.rmi.transport.Transport$1.run(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Unknown Source)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(Unknown Source)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(Unknown Source)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

### Can you reproduce the bug?

Always

### Reproduction Steps

_No response_

### Version

2.1.0

### What Atlas App Services are you using?

Local Database only

### Are you using encryption?

No

### Platform OS and version(s)

Android

### Build environment

Android Studio version: Koala | 2024.1.1 Patch 2
Android Build Tools version: 8.5.2
Gradle version: 8.7

### PR DESCRIPTION
**Pull request by KitsuneAlex**
Fixes #1825

### CODE REVIEW

**Comment by KitsuneAlex** [comment]
This seems to be the major holdback with Kotlin 2.0.20 right now.

File: packages/plugin-compiler/src/main/kotlin/io/realm/kotlin/compiler/RealmModelLoweringExtension.kt [hunk 1]
```diff
@@ -76,13 +75,13 @@ private class RealmModelLowering(private val pluginContext: IrPluginContext) : C
...
-            throw e.wrapWithCompilationException("Internal error in realm lowering", it, null)
+            throw CompilationException("Internal error in realm lowering $it", e, null)
         } catch (e: KotlinExceptionWithAttachments) {
             throw e
         } catch (e: Throwable) {
-            throw e.wrapWithCompilationException(
-                "Internal error in file lowering",
-                it,
+            throw CompilationException(
+                "Internal error in file lowering  $it",
```
Comment (pvdmde) [inline packages/plugin-compiler/src/main/kotlin/io/realm/kotlin/compiler/RealmModelLoweringExtension.kt:83]: Unlikely, but the added space may break monitoring for someone

Comment (KitsuneAlex) [reply packages/plugin-compiler/src/main/kotlin/io/realm/kotlin/compiler/RealmModelLoweringExtension.kt:83]: Oh i didn't catch that! Fixing right when i have time later today :)

**Comment by KitsuneAlex** [comment]
Doneso :)

**Comment by pvdmde** [comment]
How to ping Realm to actually merge this and release as a hotfix?

It's not so great not to support the latest stable release..

**Comment by KitsuneAlex** [comment]
Also tagging @fealebenpae @tgoyne @kraenhansen and @nicola-cab because this specific repository seems a little inactive atm and i really need this fix at work if possible.

**Comment by nirinchev** [comment]
Hey, apologies for the delays here - it's a period of multiple vacations as well as parental leaves, which has left the Kotlin team very understaffed. We expect folks to start coming back from their leaves mid-September and we'll try to prioritize this PR.

**Comment by KitsuneAlex** [comment]
> Hey, apologies for the delays here - it's a period of multiple vacations as well as parental leaves, which has left the Kotlin team very understaffed. We expect folks to start coming back from their leaves mid-September and we'll try to prioritize this PR.

Thank you very very much. And sorry if i came off as spammy here!

**Comment by nirinchev** [comment]
Not at all, it's a fair ask!

**Comment by AlexKatonaVisma** [comment]
Any news?, its a blocker for some companies to upgrading kotlin.

**Comment by nirinchev** [comment]
@nhachicha that's the PR I mentioned yesterday. Can you take a look and if all looks good, we can do a release with the latest Core fixes + this.

**Comment by nhachicha** [comment]
Thanks for the PR @KitsuneAlex some updates:
While this PR address some aspects, JetBrains introduced more compiler changes:
- Removing IrTypeBase https://github.com/JetBrains/kotlin/commit/c7e2cd8b1a29a55186d75a132778d516975c68f1
- IrFactory: https://github.com/JetBrains/kotlin/commit/7f6562285b6b228e435550e1a0a5a12fbab9a6cb this one is more critical since it moved the `IrFactory` from an interface into a class which introduces a binary incompatibility between libraries shipped with kotlin 2.0.0|10 and 2.0.20 (see my slack message for details https://kotlinlang.slack.com/archives/C7L3JB43G/p1726064330074609?thread_ts=1718831903.163399&cid=C7L3JB43G )  

I'm grouping potential fixes in https://github.com/realm/realm-kotlin/pull/1830 if you want to follow the progress

**Pull request closed without merging**